### PR TITLE
feat(TextEditor): add typography support

### PR DESCRIPTION
**Pull request by TomasSlama**
[FX-2294]

### Description

Use Typography styles in the Text Editor component in Picasso. Changes don't include styles for lists.

### How to test

- play with the Editor, try to paste some text, all text expect lists should follow BASE 2.0. 

### Review

- [ ] Read [CONTRIBUTING.md](https://github.com/toptal/picasso/blob/master/CONTRIBUTING.md) and [Component API principles](https://github.com/toptal/picasso/blob/master/docs/contribution/component-api.md)
- [ ] Annotate all `props` in component with documentation
- [ ] Create `examples` for component
- [ ] Ensure that deployed demo has expected results and good examples
- [ ] Ensure that tests pass by running `yarn test`
- [ ] Ensure that visuals tests pass by running `yarn test:visual`. If not - check the documentation [how to fix visual tests](https://github.com/toptal/picasso/blob/master/docs/contribution/visual-testing.md#fixing-broken-visual-tests-inside-a-pr)
- [ ] Ensure the changed/created components have not caused accessibility issues. [How to use accessibility plugin in storybook](https://github.com/toptal/picasso/blob/master/docs/contribution/accessibility.md).

<details>
<summary>PR commands</summary>
<br />

List of available commands:

- `@toptal-bot run all` - Run whole pipeline
- `@toptal-bot run build` - Check build
- `@toptal-bot run visual` - Run visual tests
- `@toptal-bot run deploy:documentation` - Deploy documentation
- `@toptal-bot run package:alpha-release` - Release alpha version

</details>

<details>
<summary>PR Review Guidelines</summary>
<br />

#### When to approve? ✅

**You are OK** with merging this PR and
1. You have no extra requests.
2. You have optional requests.
   1. Add `nit: ` to your comment. (ex. `nit: I'd rename this variable from makeCircle to getCircle`)

#### When to request changes? ❌

**You are not OK** with merging this PR because
1. Something is broken after the changes.
2. Acceptance criteria is not reached.
3. Code is dirty.

#### When to comment (neither ✅ nor ❌)

**You want your comments to be addressed** before merging this PR in cases like:
1. There are leftovers like unnecessary logs, comments, etc.
2. You have an opinionated comment regarding the code that requires a discussion.
3. You have questions.

#### How to handle the comments?

1. An owner of a comment is the only one who can resolve it.
2. An owner of a comment must resolve it when it's addressed.
3. A PR owner must reply with ✅ when a comment is addressed.

</details>

[FX-2294]: https://toptal-core.atlassian.net/browse/FX-2294?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

File: .storybook/tsconfig.json [hunk 1]
```diff
@@ -7,6 +7,7 @@
   "include": ["../packages", "components"],
   "exclude": [
     "../packages/*/src/**/test.@(tsx|jsx|ts|js)",
-    "../packages/*/dist-package"
+    "../packages/*/dist-package",
+    "../**/__testfixtures__"
```
Comment (TomasSlama) [inline .storybook/tsconfig.json:11]: for some reason TS started to look in new codemod files

File: packages/picasso-lab/src/TextEditor/TextEditor.tsx [hunk 2]
```diff
@@ -79,7 +80,11 @@ export const TextEditor = forwardRef<HTMLDivElement, Props>(function TextEditor(
   useHandleAutofocus(editorRef, { autofocus })
 
   return (
-    <div
+    <Typography
+      as='div'
+      variant='body'
+      color='dark-grey'
+      size='medium'
```
Comment (TomasSlama) [inline packages/picasso-lab/src/TextEditor/TextEditor.tsx:87]: this way we add basic styling for paragraphs.

File: packages/picasso-lab/src/TextEditor/TextEditor.tsx [hunk 3]
```diff
@@ -3,7 +3,8 @@ import { makeStyles, Theme } from '@material-ui/core/styles'
 import { BaseProps } from '@toptal/picasso-shared'
 import cx from 'classnames'
 import Quill from 'quill'
-import 'quill/dist/quill.snow.css'
+import { Typography } from '@toptal/picasso'
+import './quill.snow.css'
```
Comment (TomasSlama) [inline packages/picasso-lab/src/TextEditor/TextEditor.tsx:7]: I have copied styles to our own file, so I can remove typography styles.

File: packages/picasso-lab/src/TextEditor/hooks/useHandleChangeEvent.tsx [hunk 4]
```diff
@@ -18,10 +18,11 @@ const useHandleChangeEvent = (
 
     if (editor) {
       const handler: TextChangeHandler = () => {
-        const value = editor.root.innerHTML
-        const valueWithoutCursorSpan = removeCursorSpan(value)
+        const [cleanValue] = [editor.root.innerHTML]
+          .map(removeCursorSpan)
+          .map(removeClasses)
```
Comment (TomasSlama) [inline packages/picasso-lab/src/TextEditor/hooks/useHandleChangeEvent.tsx:23]: we are adding Picasso classes to HTML elements, so before we send them to value, we need to strip them.

File: packages/picasso/src/Typography/styles.ts [hunk 5]
```diff
@@ -3,47 +3,55 @@ import { PicassoProvider } from '@toptal/picasso-provider'
 
 import { toMuiVariant } from './utils'
 
+const getHeadingMedium = (theme: Theme) => ({
+  color: theme.palette.common.black,
+  fontWeight: theme.typography.fontWeights.semibold,
+  fontSize: '16px',
+  lineHeight: '24px'
+})
```
Comment (TomasSlama) [inline packages/picasso/src/Typography/styles.ts:11]: since we are adding only classes to HTML tag and not MUI Typography component, we need to have needed styles also in exported classes

File: .storybook/tsconfig.json (hunk 1 above)
Comment (OndrejTuma) [reply .storybook/tsconfig.json:11]: gaining consciousness? 🤔

File: packages/picasso-lab/src/TextEditor/TextEditor.tsx (hunk 3 above)
Comment (TomasSlama) [reply packages/picasso-lab/src/TextEditor/TextEditor.tsx:7]: We will slowly create our own styles and we will remove this CSS file.

File: packages/picasso-lab/src/TextEditor/formats/make-bold-format.ts [hunk 6]
```diff
@@ -0,0 +1,22 @@
+import { getTypographyClassName } from '@toptal/picasso/Typography/utils'
+import Quill from 'quill'
+
+const QuillBold = Quill.import('formats/bold')
+
+const makeBoldFormat = (typographyClasses: Record<string, string>) =>
```
Comment (OndrejTuma) [inline packages/picasso-lab/src/TextEditor/formats/make-bold-format.ts:6]: nit: inconsistent naming - typographyClasses vs typographyStyles

File: packages/picasso-lab/src/TextEditor/utils/remove-classes/remove-classes.ts [hunk 7]
```diff
@@ -0,0 +1,3 @@
+const removeClasses = (value: string) => value.replace(/\sclass=".+"/g, '')
```
Comment (OndrejTuma) [inline packages/picasso-lab/src/TextEditor/utils/remove-classes/remove-classes.ts:1]: This is fixed in this [PR](https://github.com/toptal/picasso/pull/2334/files)

File: packages/picasso/src/Typography/styles.ts [hunk 8]
```diff
@@ -3,47 +3,55 @@ import { PicassoProvider } from '@toptal/picasso-provider'
...
+
 // TODO: https://toptal-core.atlassian.net/browse/FX-2166
-PicassoProvider.override(({ palette, typography }) => ({
+PicassoProvider.override(theme => {
+  const { palette, typography } = theme
```
Comment (OndrejTuma) [inline packages/picasso/src/Typography/styles.ts:15]: what is the reason for this change? If I recall correctly we prefer to destruct arguments in its definition 🤔

File: packages/picasso-lab/src/TextEditor/hooks/useInitEditor.tsx [hunk 9]
```diff
@@ -37,14 +42,19 @@ const useInitEditor = (
   editorRef: EditorRefType,
   { id, placeholder }: EditorOptionsType
 ) => {
+  const typographyStyles = useTypographyStyles()
+
   useEffect(() => {
+    Quill.register(makeHeaderFormat(typographyStyles), true)
+    Quill.register(makeBoldFormat(typographyStyles), true)
+
```
Comment (OndrejTuma) [inline packages/picasso-lab/src/TextEditor/hooks/useInitEditor.tsx:50]: Doesn't have Quill some way to unregister these as a hook return function to prevent memory leak?

File: packages/picasso-lab/src/TextEditor/hooks/useHandleChangeFromController.tsx [hunk 10]
```diff
@@ -17,11 +17,12 @@ const useHandleChangeFromController = (
...
       const editor = editorRef.current
       const currentValue = editor.root.innerHTML
-
       const isValueChanged = currentValue !== controllerValue
 
       if (isValueChanged) {
-        const valueWithoutCursorSpan = removeCursorSpan(currentValue)
+        const [cleanValue] = [currentValue]
+          .map(removeCursorSpan)
+          .map(removeClasses)
```
Comment (OndrejTuma) [inline packages/picasso-lab/src/TextEditor/hooks/useHandleChangeFromController.tsx:25]: like how you did it 🙂

File: packages/picasso-lab/src/TextEditor/formats/make-header-format.ts [hunk 11]
```diff
@@ -0,0 +1,26 @@
...
+    static create() {
+      // we only support h3, so every other heading will be transformed to it
+      const node = super.create(3)
+
+      node.setAttribute(
+        'class',
+        getTypographyClassName(typographyClasses, {
+          variant: 'heading',
+          size: 'medium'
+        })
```
Comment (dmaklygin1) [inline packages/picasso-lab/src/TextEditor/formats/make-header-format.ts:19]: could you please determine, whether we can call `useTypographyStyles` right here to generate those styles?

File: packages/picasso/src/Typography/styles.ts (hunk 8 above)
Comment (TomasSlama) [reply packages/picasso/src/Typography/styles.ts:15]: @OndrejTuma yeah, we do, but I need to pass theme to new function `getHeadingMedium`, maybe I can just pass `palette` and `typography` to the function, this approach seemed clearer to me. What do you think?

File: packages/picasso-lab/src/TextEditor/hooks/useInitEditor.tsx (hunk 9 above)
Comment (TomasSlama) [reply packages/picasso-lab/src/TextEditor/hooks/useInitEditor.tsx:50]: nope https://quilljs.com/docs/api/#register

Comment (TomasSlama) [reply packages/picasso-lab/src/TextEditor/hooks/useInitEditor.tsx:50]: > For formats specifically there is a shorthand to just pass in the format directly and the path will be autogenerated. Will overwrite existing definitions with the same path.

File: packages/picasso-lab/src/TextEditor/formats/make-header-format.ts (hunk 11 above)
Comment (TomasSlama) [reply packages/picasso-lab/src/TextEditor/formats/make-header-format.ts:19]: @dmaklygin1 `useTypographyStyles` is called in `useInitEditor` and it just passes object with classes to `make-header-format/make-bold-format`
image

File: packages/picasso-lab/src/TextEditor/formats/make-bold-format.ts (hunk 6 above)
Comment (TomasSlama) [reply packages/picasso-lab/src/TextEditor/formats/make-bold-format.ts:6]: ✅
